In [2]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

In [3]:
def display_missing_data_info(dataframe, ascending=False):
    missing_values_count = dataframe.isnull().sum()
    missing_values_percentage = (dataframe.isnull().mean() * 100)
    missing_data = pd.DataFrame({
        'Missing Values': missing_values_count,
        'Percentage (%)': missing_values_percentage
    })
    
    missing_data = missing_data[missing_data['Missing Values'] > 0]
    
    sorted_missing_data = missing_data.sort_values(by='Missing Values', ascending=ascending)

    print(sorted_missing_data)
    return sorted_missing_data
    

In [4]:
bureau = pd.read_csv('../data/dseb63_bureau.csv')

In [5]:
bureau

,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,SK_ID_CURR
0,5714462,Closed,currency 1,-497,0,-153.0,-153.0,NaN,0,91323.0,0.0,NaN,0.0,Consumer credit,-131,NaN,254629
1,5714463,Active,currency 1,-208,0,1075.0,NaN,NaN,0,225000.0,171342.0,NaN,0.0,Credit card,-20,NaN,254629
2,5714464,Active,currency 1,-203,0,528.0,NaN,NaN,0,464323.5,NaN,NaN,0.0,Consumer credit,-16,NaN,254629
3,5714465,Active,currency 1,-203,0,NaN,NaN,NaN,0,90000.0,NaN,NaN,0.0,Credit card,-16,NaN,254629
4,5714466,Active,currency 1,-629,0,1197.0,NaN,77674.5,0,2700000.0,NaN,NaN,0.0,Consumer credit,-21,NaN,254629
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1465320,5052629,Closed,currency 1,-532,0,45.0,-227.0,NaN,0,450000.0,0.0,0.0,0.0,Consumer credit,-31,NaN,66820
1465321,5054382,Closed,currency 1,-381,0,-199.0,-199.0,0.0,0,19800.0,0.0,0.0,0.0,Consumer credit,-199,0.0,166221
1465322,5043236,Active,currency 1,-1441,0,2197.0,NaN,NaN,0,1800000.0,1076809.5,0.0,0.0,Mortgage,-2,NaN,118290
1465323,5053303,Closed,currency 1,-375,0,-41.0,-19.0,NaN,0,175054.5,NaN,NaN,0.0,Consumer credit,-4,17140.5,95296


In [ ]:
appl_train = pd.read_csv('../data/dseb63_application_train.csv')
bureau = pd.merge(bureau, appl_train[['SK_ID_CURR', 'TARGET']], on='SK_ID_CURR', how='left')

In [27]:
bureau['CREDIT_TYPE'].value_counts()

CREDIT_TYPE
Consumer credit                                 1069600
Credit card                                      343710
Car loan                                          23757
Mortgage                                          15607
Microloan                                          9005
Loan for business development                      1721
Another type of loan                                854
Unknown type of loan                                472
Loan for working capital replenishment              377
Real estate loan                                     23
Cash loan (non-earmarked)                            23
Loan for the purchase of equipment                   17
Loan for purchase of shares (margin lending)          4
Mobile operator loan                                  1
Interbank credit                                      1
Name: count, dtype: int64

In [28]:
others_loan = ['Another type of loan', 'Unknown type of loan', 'Loan for working capital replenishment', 'Cash loan (non-earmarked)', 'Real estate loan', 'Loan for the purchase of equipment', 'Mobile operator loan', 'Interbank credit']
mapping_others_loan= {x: 'Others' for x in others_loan}

bureau['CREDIT_TYPE'] = bureau['CREDIT_TYPE'].replace(mapping_others_loan)
bureau

/tmp/ipykernel_197616/1667981891.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bureau['CREDIT_TYPE'] = bureau['CREDIT_TYPE'].replace(mapping_others_loan)


,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,SK_ID_CURR,DURATION
0,5714462,Closed,currency 1,-497,0,-153.0,-153.0,NaN,0,91323.0,0.0,NaN,0.0,Consumer credit,-131,NaN,254629,344.0
1,5714463,Active,currency 1,-208,0,1075.0,NaN,NaN,0,225000.0,171342.0,NaN,0.0,Credit card,-20,NaN,254629,1283.0
2,5714464,Active,currency 1,-203,0,528.0,NaN,NaN,0,464323.5,NaN,NaN,0.0,Consumer credit,-16,NaN,254629,731.0
3,5714465,Active,currency 1,-203,0,365243.0,NaN,NaN,0,90000.0,NaN,NaN,0.0,Credit card,-16,NaN,254629,365446.0
4,5714466,Active,currency 1,-629,0,1197.0,NaN,77674.5,0,2700000.0,NaN,NaN,0.0,Consumer credit,-21,NaN,254629,1826.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1465320,5052629,Closed,currency 1,-532,0,45.0,-227.0,NaN,0,450000.0,0.0,0.0,0.0,Consumer credit,-31,NaN,66820,577.0
1465321,5054382,Closed,currency 1,-381,0,-199.0,-199.0,0.0,0,19800.0,0.0,0.0,0.0,Consumer credit,-199,0.0,166221,182.0
1465322,5043236,Active,currency 1,-1441,0,2197.0,NaN,NaN,0,1800000.0,1076809.5,0.0,0.0,Mortgage,-2,NaN,118290,3638.0
1465323,5053303,Closed,currency 1,-375,0,-41.0,-19.0,NaN,0,175054.5,NaN,NaN,0.0,Consumer credit,-4,17140.5,95296,334.0


In [47]:
bureau[bureau['TARGET'] == 1].head(20)

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET,DURATION
129,338464,5714626,Closed,currency 1,-2655,0,-2319.0,-2348.0,0.000,0,20165.400,0.00,0.000,0.0,Consumer credit,-831,NaN,1.0,336.0
130,338464,5714627,Closed,currency 1,-2539,0,-1427.0,-651.0,0.000,0,45000.000,0.00,0.000,0.0,Credit card,-651,NaN,1.0,1112.0
131,338464,5714628,Closed,currency 1,-2343,0,-2008.0,-2219.0,0.000,0,19192.500,0.00,0.000,0.0,Consumer credit,-2219,NaN,1.0,335.0
132,338464,5714629,Closed,currency 1,-857,0,-523.0,-582.0,3285.000,0,33853.500,0.00,0.000,0.0,Consumer credit,-582,NaN,1.0,334.0
133,338464,5714630,Active,currency 1,-857,0,765.0,NaN,1704.780,0,45000.000,38294.37,6705.630,0.0,Credit card,-675,NaN,1.0,1622.0
134,338464,5714631,Active,currency 1,-1131,0,-27.0,NaN,NaN,0,270000.000,0.00,0.000,0.0,Credit card,-17,NaN,1.0,1104.0
135,338464,5714633,Closed,currency 1,-1706,0,-976.0,-1260.0,NaN,0,247500.000,0.00,0.000,0.0,Consumer credit,-538,NaN,1.0,730.0
136,338464,5714634,Active,currency 1,-140,0,365243.0,NaN,0.000,0,144000.000,143019.00,980.280,0.0,Credit card,-22,NaN,1.0,365383.0
197,280788,5714715,Active,currency 1,-221,0,875.0,NaN,NaN,0,220675.500,NaN,NaN,0.0,Consumer credit,-30,NaN,1.0,1096.0
198,280788,5714716,Active,currency 1,-221,0,365243.0,NaN,NaN,0,90000.000,NaN,NaN,0.0,Credit card,-9,NaN,1.0,365464.0


In [6]:
bureau['CREDIT_TYPE'].value_counts()

CREDIT_TYPE
Consumer credit                                 1251615
Credit card                                      402195
Car loan                                          27690
Mortgage                                          18391
Microloan                                         12413
Loan for business development                      1975
Another type of loan                               1017
Unknown type of loan                                555
Loan for working capital replenishment              469
Cash loan (non-earmarked)                            56
Real estate loan                                     27
Loan for the purchase of equipment                   19
Loan for purchase of shares (margin lending)          4
Mobile operator loan                                  1
Interbank credit                                      1
Name: count, dtype: int64

In [7]:
bureau['CREDIT_CURRENCY'].value_counts()

CREDIT_CURRENCY
currency 1    1715020
currency 2       1224
currency 3        174
currency 4         10
Name: count, dtype: int64

In [8]:
bureau_not_null = bureau[bureau['AMT_ANNUITY'].notnull()]
bureau_not_null

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET
768,380361,5715448,Active,currency 1,-820,0,31069.0,NaN,NaN,0,67500.0,0.0,67500.0,0.0,Credit card,-183,0.0,0.0
769,380361,5715449,Active,currency 1,-357,0,1119.0,NaN,NaN,0,45000.0,0.0,45000.0,0.0,Credit card,-130,2691.0,0.0
770,380361,5715451,Closed,currency 1,-917,0,-187.0,-759.0,NaN,0,74439.0,0.0,0.0,0.0,Consumer credit,-748,0.0,0.0
771,380361,5715452,Closed,currency 1,-993,0,31039.0,-831.0,NaN,0,315000.0,0.0,0.0,0.0,Credit card,-818,0.0,0.0
772,380361,5715453,Closed,currency 1,-1146,0,681.0,-780.0,NaN,0,2025000.0,0.0,0.0,0.0,Consumer credit,-769,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1716285,435426,5053678,Closed,currency 1,-1147,0,-416.0,-849.0,0.0,0,205488.0,0.0,0.0,0.0,Consumer credit,-835,94378.5,0.0
1716286,377669,5053682,Closed,currency 1,-2215,0,-1843.0,-1653.0,NaN,0,58027.5,0.0,0.0,0.0,Consumer credit,-1642,0.0,0.0
1716287,431028,5053695,Closed,currency 1,-586,0,-191.0,-177.0,NaN,0,1350000.0,0.0,NaN,0.0,Consumer credit,-176,0.0,0.0
1716290,363575,5053740,Closed,currency 1,-1234,0,-503.0,-502.0,NaN,0,900000.0,0.0,0.0,0.0,Consumer credit,-461,58554.0,0.0


In [9]:
bureau_diff_currency = bureau[bureau['CREDIT_CURRENCY'] != 'currency 1']
bureau_diff_currency

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET
1436,108012,5716224,Closed,currency 2,-2889,0,-1063.0,-1061.0,NaN,0,2790000.0,0.0,0.0,0.0,Consumer credit,-1059,NaN,0.0
1475,430110,5716274,Closed,currency 2,-1974,0,-1609.0,-1758.0,NaN,0,177444.0,0.0,NaN,0.0,Consumer credit,-1756,NaN,0.0
2071,316711,5717036,Closed,currency 2,-2194,0,-1830.0,-1815.0,NaN,0,1181286.0,NaN,NaN,0.0,Car loan,-1815,NaN,0.0
6835,253537,5722696,Closed,currency 2,-2817,0,-1723.0,-2773.0,NaN,0,4737420.0,0.0,0.0,0.0,Car loan,-1510,NaN,0.0
8303,270344,5724552,Closed,currency 4,-1791,0,35.0,-1365.0,0.0,0,973350.0,0.0,0.0,0.0,Consumer credit,-1365,NaN,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1705093,258972,6843229,Closed,currency 2,-2672,0,-1576.0,-2479.0,NaN,0,1081125.0,0.0,NaN,0.0,Car loan,-2479,NaN,0.0
1705358,218305,5025399,Closed,currency 2,-1989,0,-1443.0,-1686.0,NaN,0,964503.0,0.0,NaN,0.0,Consumer credit,-1684,NaN,0.0
1713847,269375,5079061,Closed,currency 2,-1832,0,-1467.0,-1462.0,NaN,0,156519.0,0.0,NaN,0.0,Consumer credit,-1460,NaN,0.0
1715132,333280,5088164,Closed,currency 2,-2523,0,-696.0,-2096.0,NaN,0,3953709.0,0.0,0.0,0.0,Car loan,-590,3378969.0,0.0


In [10]:
bureau[bureau['CREDIT_ACTIVE'] == 'Bad debt']


,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET
7778,436084,5723864,Bad debt,currency 1,-986,0,NaN,NaN,184500.000,0,184500.00,0.000,0.000,0.00,Credit card,-22,NaN,1.0
30959,396333,5801236,Bad debt,currency 1,-2627,0,-1529.0,-1502.0,127866.915,0,225000.00,115349.715,-25349.715,0.00,Credit card,-1502,NaN,0.0
69946,386859,5246470,Bad debt,currency 1,-2386,26,-2296.0,NaN,53532.000,1,58059.00,NaN,NaN,53532.00,Consumer credit,-2296,NaN,1.0
295159,231185,5173404,Bad debt,currency 1,-2740,1761,-2558.0,NaN,2137.500,0,14296.50,7717.500,0.000,7717.50,Consumer credit,-888,NaN,0.0
544624,446651,6439586,Bad debt,currency 1,-2830,0,-2434.0,-1781.0,21741.615,0,58567.68,0.000,0.000,0.00,Consumer credit,-1781,NaN,0.0
546312,232061,6441729,Bad debt,currency 1,-2899,0,-1773.0,-1434.0,242637.480,0,365670.72,0.000,0.000,0.00,Consumer credit,-1434,NaN,0.0
550115,243877,6446445,Bad debt,currency 1,-2493,366,-898.0,NaN,NaN,0,162000.00,NaN,NaN,17541.00,Credit card,-1449,NaN,0.0
684142,273612,5309530,Bad debt,currency 1,-2112,0,-1900.0,-1550.0,40871.880,0,22510.89,0.000,0.000,0.00,Consumer credit,-1550,NaN,0.0
711352,186360,5499851,Bad debt,currency 1,-1218,366,-852.0,NaN,NaN,0,121275.00,0.000,NaN,0.00,Consumer credit,-1035,NaN,0.0
858406,158069,6039562,Bad debt,currency 1,-1683,366,-862.0,NaN,NaN,0,528165.00,NaN,NaN,503392.50,Consumer credit,-1217,NaN,0.0


In [ ]:
not_end =  bureau[~(bureau['DAYS_ENDDATE_FACT']<0)]
not_end

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET
1,215354,5714463,Active,currency 1,-208,0,1075.0,NaN,NaN,0,225000.0,171342.00,NaN,0.0,Credit card,-20,NaN,0.0
2,215354,5714464,Active,currency 1,-203,0,528.0,NaN,NaN,0,464323.5,NaN,NaN,0.0,Consumer credit,-16,NaN,0.0
3,215354,5714465,Active,currency 1,-203,0,NaN,NaN,NaN,0,90000.0,NaN,NaN,0.0,Credit card,-16,NaN,0.0
4,215354,5714466,Active,currency 1,-629,0,1197.0,NaN,77674.5,0,2700000.0,NaN,NaN,0.0,Consumer credit,-21,NaN,0.0
5,215354,5714467,Active,currency 1,-273,0,27460.0,NaN,0.0,0,180000.0,71017.38,108982.62,0.0,Credit card,-31,NaN,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1716404,166279,5057599,Active,currency 1,-99,0,997.0,NaN,0.0,0,450000.0,432787.50,0.00,0.0,Consumer credit,-7,NaN,0.0
1716410,280228,5057642,Active,currency 1,-575,0,813.0,NaN,NaN,0,67500.0,0.00,0.00,0.0,Credit card,-46,NaN,0.0
1716412,288038,5057657,Active,currency 1,-541,0,7.0,NaN,NaN,0,225000.0,10971.00,0.00,0.0,Consumer credit,-11,NaN,0.0
1716417,433007,5057703,Active,currency 1,-740,0,31128.0,NaN,NaN,0,67500.0,2466.00,NaN,0.0,Credit card,-433,NaN,0.0


In [12]:
bureau['CREDIT_ACTIVE'].value_counts()

CREDIT_ACTIVE
Closed      1079273
Active       630607
Sold           6527
Bad debt         21
Name: count, dtype: int64

In [13]:
bureau[bureau['SK_ID_BUREAU'] == 5041336]

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET
1709370,101874,5041336,Active,currency 1,-1579,0,-1274.0,NaN,30519.855,0,142317.0,0.0,0.0,0.0,Consumer credit,-111,NaN,1.0


ADD Feature

## General
- Count record
- SUM overdue
- Count currency
- Sum credit
- Sum debt
- Sum overdue
- Count credit_type
- sum everything :))
- Credit sum on Car loan
- Credit sum on Consumer credit
- Credit sum on Credit Card
- Credit sum on Mortgage
- Credit sum on Other

In [ ]:
gb_bureau_general = bureau.groupby('SK_ID_CURR').agg({'SK_ID_BUREAU': 'count', 'AMT_CREDIT_SUM': 'sum', 'AMT_CREDIT_SUM_DEBT': 'sum', 'AMT_CREDIT_SUM_OVERDUE': 'sum', 'AMT_ANNUITY': 'sum', 'CREDIT_CURRENCY': 'nunique', 'CREDIT_TYPE': 'nunique', 'CREDIT_ACTIVE': 'nunique' }).reset_index()

sum_on_car_loan = bureau[bureau['CREDIT_TYPE'] == 'Car loan'].groupby('SK_ID_CURR').agg({'AMT_CREDIT_SUM': 'sum', 'AMT_CREDIT_SUM_DEBT': 'sum', 'AMT_CREDIT_SUM_OVERDUE': 'sum'}).reset_index()
sum_on_car_loan.rename(columns={'AMT_CREDIT_SUM':'SUM_CREDIT_CAR_LOAN', 'AMT_CREDIT_SUM_DEBT':'SUM_DEBT_CAR_LOAN', 'AMT_CREDIT_SUM_OVERDUE':'SUM_OVERDUE_CAR_LOAN' }, inplace=True)

sum_on_credit_card = bureau[bureau['CREDIT_TYPE'] == 'Credit card'].groupby('SK_ID_CURR').agg({'AMT_CREDIT_SUM': 'sum', 'AMT_CREDIT_SUM_DEBT': 'sum', 'AMT_CREDIT_SUM_OVERDUE': 'sum'}).reset_index()
sum_on_credit_card.rename(columns={'AMT_CREDIT_SUM':'SUM_CREDIT_CARD', 'AMT_CREDIT_SUM_DEBT':'SUM_DEBT_CARD', 'AMT_CREDIT_SUM_OVERDUE':'SUM_OVERDUE_CARD' }, inplace=True)

sum_on_mortgage = bureau[bureau['CREDIT_TYPE'] == 'Mortgage'].groupby('SK_ID_CURR').agg({'AMT_CREDIT_SUM': 'sum', 'AMT_CREDIT_SUM_DEBT': 'sum', 'AMT_CREDIT_SUM_OVERDUE': 'sum'}).reset_index()
sum_on_mortgage.rename(columns={'AMT_CREDIT_SUM':'SUM_MORTGAGE', 'AMT_CREDIT_SUM_DEBT':'SUM_DEBT_MORTGAGE', 'AMT_CREDIT_SUM_OVERDUE':'SUM_OVERDUE_MORTGAGE' }, inplace=True)

sum_on_consumer_credit = bureau[bureau['CREDIT_TYPE'] == 'Consumer credit'].groupby('SK_ID_CURR').agg({'AMT_CREDIT_SUM': 'sum', 'AMT_CREDIT_SUM_DEBT': 'sum', 'AMT_CREDIT_SUM_OVERDUE': 'sum'}).reset_index()
sum_on_consumer_credit.rename(columns={'AMT_CREDIT_SUM':'SUM_CONSUMER_CREDIT', 'AMT_CREDIT_SUM_DEBT':'SUM_DEBT_CONSUMER_CREDIT', 'AMT_CREDIT_SUM_OVERDUE':'SUM_OVERDUE_CONSUMER_CREDIT' }, inplace=True)

sum_on_other = bureau[bureau['CREDIT_TYPE']== 'Others'].groupby('SK_ID_CURR').agg({'AMT_CREDIT_SUM': 'sum'})
sum_on_other.rename(columns={'AMT_CREDIT_SUM':'SUM_OTHER'}, inplace=True)

In [18]:
sum_on_car_loan

,SK_ID_CURR,SUM_CREDIT_CAR_LOAN,SUM_DEBT_CAR_LOAN,SUM_OVERDUE_CAR_LOAN
0,8,2351439.0,0.00,0.0
1,18,2608200.0,1685826.00,0.0
2,47,3972388.5,0.00,0.0
3,54,4252050.0,0.00,0.0
4,69,1675080.0,0.00,0.0
...,...,...,...,...
19570,307347,1620000.0,0.00,0.0
19571,307365,2700000.0,2644735.32,0.0
19572,307393,1395000.0,0.00,0.0
19573,307414,1777500.0,0.00,0.0


In [19]:
pd.merge(gb_bureau_general, sum_on_car_loan, on='SK_ID_CURR', how='left')

,SK_ID_CURR,SK_ID_BUREAU,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_OVERDUE,AMT_ANNUITY,CREDIT_CURRENCY,CREDIT_TYPE,CREDIT_ACTIVE,SUM_CREDIT_CAR_LOAN,SUM_DEBT_CAR_LOAN,SUM_OVERDUE_CAR_LOAN
0,0,1,85513.50,77566.5,0.0,0.000,1,1,1,NaN,NaN,NaN
1,1,1,28350.00,0.0,0.0,0.000,1,1,1,NaN,NaN,NaN
2,2,6,254341.08,96417.0,0.0,0.000,1,1,2,NaN,NaN,NaN
3,3,7,483952.50,198769.5,0.0,0.000,1,2,2,NaN,NaN,NaN
4,4,6,2822544.00,1254505.5,0.0,0.000,1,2,2,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
263486,307504,4,866153.25,119767.5,0.0,0.000,1,1,2,NaN,NaN,NaN
263487,307505,2,3618000.00,0.0,0.0,180927.000,1,2,1,NaN,NaN,NaN
263488,307506,4,259263.00,0.0,0.0,0.000,1,1,1,NaN,NaN,NaN
263489,307508,7,5506306.83,251748.0,0.0,25476.255,1,3,2,NaN,NaN,NaN


In [13]:
gb_bureau_general.head()

,SK_ID_CURR,SK_ID_BUREAU,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_OVERDUE,AMT_ANNUITY,CREDIT_CURRENCY,CREDIT_TYPE,CREDIT_ACTIVE
0,0,1,85513.50,77566.5,0.0,0.0,1,1,1
1,1,1,28350.00,0.0,0.0,0.0,1,1,1
2,2,6,254341.08,96417.0,0.0,0.0,1,1,2
3,3,7,483952.50,198769.5,0.0,0.0,1,2,2
4,4,6,2822544.00,1254505.5,0.0,0.0,1,2,2


In [20]:
gb_bureau_general = pd.merge(gb_bureau_general, sum_on_car_loan, on='SK_ID_CURR', how='left')
gb_bureau_general = pd.merge(gb_bureau_general, sum_on_credit_card, on='SK_ID_CURR', how='left')
gb_bureau_general = pd.merge(gb_bureau_general, sum_on_mortgage, on='SK_ID_CURR', how='left')
gb_bureau_general = pd.merge(gb_bureau_general, sum_on_consumer_credit, on='SK_ID_CURR', how='left')
gb_bureau_general = pd.merge(gb_bureau_general, sum_on_other, on='SK_ID_CURR', how='left')
gb_bureau_general.head()

,SK_ID_CURR,SK_ID_BUREAU,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_OVERDUE,AMT_ANNUITY,CREDIT_CURRENCY,CREDIT_TYPE,CREDIT_ACTIVE,SUM_CREDIT_CAR_LOAN,...,SUM_CREDIT_CARD,SUM_DEBT_CARD,SUM_OVERDUE_CARD,SUM_MORTGAGE,SUM_DEBT_MORTGAGE,SUM_OVERDUE_MORTGAGE,SUM_CONSUMER_CREDIT,SUM_DEBT_CONSUMER_CREDIT,SUM_OVERDUE_CONSUMER_CREDIT,SUM_OTHER
0,0,1,85513.50,77566.5,0.0,0.0,1,1,1,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,85513.50,77566.5,0.0,NaN
1,1,1,28350.00,0.0,0.0,0.0,1,1,1,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,28350.00,0.0,0.0,NaN
2,2,6,254341.08,96417.0,0.0,0.0,1,1,2,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,254341.08,96417.0,0.0,NaN
3,3,7,483952.50,198769.5,0.0,0.0,1,2,2,NaN,...,148500.0,157698.0,0.0,NaN,NaN,NaN,335452.50,41071.5,0.0,NaN
4,4,6,2822544.00,1254505.5,0.0,0.0,1,2,2,NaN,...,270000.0,261477.0,0.0,NaN,NaN,NaN,2552544.00,993028.5,0.0,NaN


In [21]:
gb_bureau_general.to_csv('../temp/dseb63_bureau_general_v1.csv', index=False)

In [14]:
bureau[bureau['DAYS_CREDIT_ENDDATE'].isnull()]

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET
3,215354,5714465,Active,currency 1,-203,0,NaN,NaN,NaN,0,90000.0,NaN,NaN,0.0,Credit card,-16,NaN,0.0
17,238881,5714486,Active,currency 1,-381,0,NaN,NaN,NaN,0,450000.0,520920.00,NaN,0.0,Consumer credit,-4,NaN,0.0
20,238881,5714489,Active,currency 1,-392,0,NaN,NaN,0.0,0,252000.0,23679.00,228320.10,0.0,Credit card,-22,NaN,0.0
32,136226,5714503,Closed,currency 1,-559,0,NaN,-355.0,0.0,0,110250.0,0.00,0.00,0.0,Credit card,-351,NaN,0.0
43,435112,5714518,Active,currency 1,-1483,0,NaN,NaN,NaN,0,200250.0,0.00,NaN,0.0,Credit card,-848,NaN,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1716324,104893,5055269,Active,currency 1,-51,0,NaN,NaN,0.0,0,103500.0,92882.25,10617.75,0.0,Credit card,-13,NaN,0.0
1716351,169747,5055520,Active,currency 1,-307,0,NaN,NaN,0.0,0,112500.0,0.00,112500.00,0.0,Credit card,-24,NaN,0.0
1716384,195589,5057408,Active,currency 1,-195,0,NaN,NaN,NaN,0,90000.0,NaN,NaN,0.0,Credit card,-5,NaN,0.0
1716396,181001,5057529,Closed,currency 1,-339,0,NaN,-163.0,NaN,0,1057500.0,NaN,NaN,0.0,Consumer credit,-163,NaN,0.0


In [15]:
bureau_limit_duration = bureau[~bureau['DAYS_CREDIT_ENDDATE'].isnull()]
bureau_limit_duration

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET
0,215354,5714462,Closed,currency 1,-497,0,-153.0,-153.0,NaN,0,91323.00,0.00,NaN,0.0,Consumer credit,-131,NaN,0.0
1,215354,5714463,Active,currency 1,-208,0,1075.0,NaN,NaN,0,225000.00,171342.00,NaN,0.0,Credit card,-20,NaN,0.0
2,215354,5714464,Active,currency 1,-203,0,528.0,NaN,NaN,0,464323.50,NaN,NaN,0.0,Consumer credit,-16,NaN,0.0
4,215354,5714466,Active,currency 1,-629,0,1197.0,NaN,77674.5,0,2700000.00,NaN,NaN,0.0,Consumer credit,-21,NaN,0.0
5,215354,5714467,Active,currency 1,-273,0,27460.0,NaN,0.0,0,180000.00,71017.38,108982.62,0.0,Credit card,-31,NaN,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1716422,375755,5057742,Closed,currency 1,-2648,0,31129.0,-189.0,NaN,0,202500.00,0.00,NaN,0.0,Credit card,-109,NaN,0.0
1716423,259355,5057750,Active,currency 1,-44,0,-30.0,NaN,0.0,0,11250.00,11250.00,0.00,0.0,Microloan,-19,NaN,1.0
1716424,100044,5057754,Closed,currency 1,-2648,0,-2433.0,-2493.0,5476.5,0,38130.84,0.00,0.00,0.0,Consumer credit,-2493,NaN,0.0
1716425,100044,5057762,Closed,currency 1,-1809,0,-1628.0,-970.0,NaN,0,15570.00,NaN,NaN,0.0,Consumer credit,-967,NaN,0.0


In [16]:
bureau_limit_duration[bureau_limit_duration['DAYS_CREDIT_ENDDATE'] > 20000]

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET
5,215354,5714467,Active,currency 1,-273,0,27460.0,NaN,0.0,0,180000.0,71017.38,108982.62,0.0,Credit card,-31,NaN,0.0
141,444864,5714639,Closed,currency 1,-2030,0,31111.0,-1750.0,NaN,0,85050.0,0.00,0.00,0.0,Credit card,-555,NaN,0.0
180,286014,5714690,Active,currency 1,-137,0,27460.0,NaN,0.0,0,22500.0,18765.63,76.59,0.0,Credit card,-31,NaN,0.0
241,228777,5714776,Active,currency 1,-1144,0,31081.0,NaN,NaN,0,225000.0,0.00,0.00,0.0,Credit card,-41,NaN,0.0
393,221368,5714983,Active,currency 1,-64,0,27460.0,NaN,0.0,0,22500.0,0.00,22500.00,0.0,Credit card,-17,NaN,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1716322,392435,5055257,Active,currency 1,-151,0,31092.0,NaN,NaN,0,40500.0,32827.50,NaN,0.0,Credit card,-46,NaN,1.0
1716359,366785,5055579,Active,currency 1,-221,0,31090.0,NaN,NaN,0,45000.0,22941.00,NaN,0.0,Credit card,-221,NaN,1.0
1716379,250798,5057380,Active,currency 1,-471,0,31129.0,NaN,NaN,0,135000.0,6588.00,NaN,0.0,Credit card,-432,NaN,0.0
1716417,433007,5057703,Active,currency 1,-740,0,31128.0,NaN,NaN,0,67500.0,2466.00,NaN,0.0,Credit card,-433,NaN,0.0


Drop all the Duration < 0

In [24]:
bureau['DAYS_CREDIT_ENDDATE'] = bureau['DAYS_CREDIT_ENDDATE'].fillna(365243)
bureau['DURATION'] = bureau['DAYS_CREDIT_ENDDATE'] - bureau['DAYS_CREDIT']

In [25]:
(bureau['DURATION']<0).sum()

153

In [26]:
bureau = bureau[bureau['DURATION'] >= 0]

CREDIT_TYPE
Consumer credit                                 1251603
Credit card                                      402052
Car loan                                          27690
Mortgage                                          18391
Microloan                                         12412
Loan for business development                      1975
Another type of loan                               1009
Unknown type of loan                                555
Loan for working capital replenishment              469
Cash loan (non-earmarked)                            56
Real estate loan                                     27
Loan for the purchase of equipment                   19
Loan for purchase of shares (margin lending)          4
Mobile operator loan                                  1
Interbank credit                                      1
Name: count, dtype: int64

In [44]:
bureau_active = bureau[bureau['CREDIT_ACTIVE'] == 'Active']
bureau_active

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET,DURATION
1,215354,5714463,Active,currency 1,-208,0,1075.0,NaN,NaN,0,225000.0,171342.00,NaN,0.0,Credit card,-20,NaN,0.0,1283.0
2,215354,5714464,Active,currency 1,-203,0,528.0,NaN,NaN,0,464323.5,NaN,NaN,0.0,Consumer credit,-16,NaN,0.0,731.0
3,215354,5714465,Active,currency 1,-203,0,365243.0,NaN,NaN,0,90000.0,NaN,NaN,0.0,Credit card,-16,NaN,0.0,365446.0
4,215354,5714466,Active,currency 1,-629,0,1197.0,NaN,77674.5,0,2700000.0,NaN,NaN,0.0,Consumer credit,-21,NaN,0.0,1826.0
5,215354,5714467,Active,currency 1,-273,0,27460.0,NaN,0.0,0,180000.0,71017.38,108982.62,0.0,Credit card,-31,NaN,0.0,27733.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1716404,166279,5057599,Active,currency 1,-99,0,997.0,NaN,0.0,0,450000.0,432787.50,0.00,0.0,Consumer credit,-7,NaN,0.0,1096.0
1716410,280228,5057642,Active,currency 1,-575,0,813.0,NaN,NaN,0,67500.0,0.00,0.00,0.0,Credit card,-46,NaN,0.0,1388.0
1716412,288038,5057657,Active,currency 1,-541,0,7.0,NaN,NaN,0,225000.0,10971.00,0.00,0.0,Consumer credit,-11,NaN,0.0,548.0
1716417,433007,5057703,Active,currency 1,-740,0,31128.0,NaN,NaN,0,67500.0,2466.00,NaN,0.0,Credit card,-433,NaN,0.0,31868.0


### Feature for Active
- Count active
- Count active in 30 days
- Count active in 90 days
- Count active in 180 days
- Count active in 365 days
- Count active credit-type
- Count life-time duration
- Sum annuity

In [24]:
bureau_closed = bureau[bureau['CREDIT_ACTIVE'] == 'Closed']
bureau_closed

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET,DURATION
0,215354,5714462,Closed,currency 1,-497,0,-153.0,-153.0,NaN,0,91323.00,0.0,NaN,0.0,Consumer credit,-131,NaN,0.0,344.0
7,162297,5714469,Closed,currency 1,-1896,0,-1684.0,-1710.0,14985.0,0,76878.45,0.0,0.0,0.0,Consumer credit,-1710,NaN,0.0,212.0
8,162297,5714470,Closed,currency 1,-1146,0,-811.0,-840.0,0.0,0,103007.70,0.0,0.0,0.0,Consumer credit,-840,NaN,0.0,335.0
11,162297,5714473,Closed,currency 1,-2456,0,-629.0,-825.0,NaN,0,675000.00,0.0,0.0,0.0,Consumer credit,-706,NaN,0.0,1827.0
14,238881,5714482,Closed,currency 1,-318,0,-187.0,-187.0,NaN,0,0.00,0.0,0.0,0.0,Credit card,-185,NaN,0.0,131.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1716422,375755,5057742,Closed,currency 1,-2648,0,31129.0,-189.0,NaN,0,202500.00,0.0,NaN,0.0,Credit card,-109,NaN,0.0,33777.0
1716424,100044,5057754,Closed,currency 1,-2648,0,-2433.0,-2493.0,5476.5,0,38130.84,0.0,0.0,0.0,Consumer credit,-2493,NaN,0.0,215.0
1716425,100044,5057762,Closed,currency 1,-1809,0,-1628.0,-970.0,NaN,0,15570.00,NaN,NaN,0.0,Consumer credit,-967,NaN,0.0,181.0
1716426,246829,5057770,Closed,currency 1,-1878,0,-1513.0,-1513.0,NaN,0,36000.00,0.0,0.0,0.0,Consumer credit,-1508,NaN,0.0,365.0


In [33]:
bureau_closed = bureau_closed[bureau_closed['DAYS_ENDDATE_FACT'] >= -16000]

In [34]:
bureau_closed['DAYS_ENDDATE_FACT'].value_counts()

DAYS_ENDDATE_FACT
-329.0     811
-273.0     794
-301.0     790
-182.0     780
-154.0     779
          ... 
-2916.0      1
-2914.0      1
-2877.0      1
-2901.0      1
-2889.0      1
Name: count, Length: 2909, dtype: int64

In [32]:
bureau_closed[bureau_closed['DAYS_ENDDATE_FACT'] == -42023]

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET,DURATION
198346,104463,6550986,Closed,currency 1,-2340,0,-1836.0,-42023.0,0.0,0,225000.0,0.0,225000.0,0.0,Credit card,-2196,NaN,0.0,504.0


In [36]:
bureau_closed['LATENCY'] =  bureau_closed['DAYS_CREDIT_ENDDATE'] - bureau_closed['DAYS_ENDDATE_FACT']
bureau_closed

C:\Users\Admin\AppData\Local\Temp\ipykernel_30796\873638872.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bureau_closed['LATENCY'] =  bureau_closed['DAYS_CREDIT_ENDDATE'] - bureau_closed['DAYS_ENDDATE_FACT']


,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET,DURATION,LATENCY
0,215354,5714462,Closed,currency 1,-497,0,-153.0,-153.0,NaN,0,91323.00,0.0,NaN,0.0,Consumer credit,-131,NaN,0.0,344.0,0.0
7,162297,5714469,Closed,currency 1,-1896,0,-1684.0,-1710.0,14985.0,0,76878.45,0.0,0.0,0.0,Consumer credit,-1710,NaN,0.0,212.0,26.0
8,162297,5714470,Closed,currency 1,-1146,0,-811.0,-840.0,0.0,0,103007.70,0.0,0.0,0.0,Consumer credit,-840,NaN,0.0,335.0,29.0
11,162297,5714473,Closed,currency 1,-2456,0,-629.0,-825.0,NaN,0,675000.00,0.0,0.0,0.0,Consumer credit,-706,NaN,0.0,1827.0,196.0
14,238881,5714482,Closed,currency 1,-318,0,-187.0,-187.0,NaN,0,0.00,0.0,0.0,0.0,Credit card,-185,NaN,0.0,131.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1716422,375755,5057742,Closed,currency 1,-2648,0,31129.0,-189.0,NaN,0,202500.00,0.0,NaN,0.0,Credit card,-109,NaN,0.0,33777.0,31318.0
1716424,100044,5057754,Closed,currency 1,-2648,0,-2433.0,-2493.0,5476.5,0,38130.84,0.0,0.0,0.0,Consumer credit,-2493,NaN,0.0,215.0,60.0
1716425,100044,5057762,Closed,currency 1,-1809,0,-1628.0,-970.0,NaN,0,15570.00,NaN,NaN,0.0,Consumer credit,-967,NaN,0.0,181.0,-658.0
1716426,246829,5057770,Closed,currency 1,-1878,0,-1513.0,-1513.0,NaN,0,36000.00,0.0,0.0,0.0,Consumer credit,-1508,NaN,0.0,365.0,0.0


### Feature for close Bureau

- Count close bureau
- Sum overdue
- Count late
- Count early
- AVG duration
- AVG early
- AVG late
- 30 Day
- 90 Day
- 180 Day
- 365 Day
- 730 Day
- Sum annuity

In [25]:
bureau_closed.dropna(subset=['DAYS_ENDDATE_FACT'], inplace=True)

C:\Users\Admin\AppData\Local\Temp\ipykernel_30796\1030630136.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bureau_closed.dropna(subset=['DAYS_ENDDATE_FACT'], inplace=True)


In [26]:
bureau_sold = bureau[bureau['CREDIT_ACTIVE'] == 'Sold']
bureau_sold

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET,DURATION
541,161678,5715165,Sold,currency 1,-1319,0,-1134.0,-75.0,133308.81,0,202500.0,0.0,0.0,0.0,Consumer credit,-71,NaN,1.0,185.0
746,147546,5715420,Sold,currency 1,-1337,0,-239.0,NaN,NaN,0,477000.0,0.0,NaN,0.0,Consumer credit,-280,NaN,0.0,1098.0
1077,113548,5715804,Sold,currency 1,-151,0,764.0,NaN,NaN,0,435973.5,NaN,NaN,0.0,Consumer credit,-24,NaN,0.0,915.0
1533,163163,5716347,Sold,currency 1,-2047,0,-1499.0,NaN,NaN,0,153405.0,NaN,NaN,0.0,Credit card,-1957,NaN,0.0,548.0
2493,351664,5717545,Sold,currency 1,-1911,0,-815.0,-134.0,42988.50,0,222750.0,NaN,NaN,0.0,Consumer credit,-22,NaN,1.0,1096.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1714457,347814,5099924,Sold,currency 1,-2095,0,-1911.0,NaN,NaN,0,23589.0,NaN,NaN,0.0,Credit card,-2057,0.0,0.0,184.0
1715449,102875,5052790,Sold,currency 1,-2327,0,-1960.0,NaN,NaN,0,174222.0,NaN,NaN,0.0,Credit card,-2113,0.0,1.0,367.0
1715602,370974,5043238,Sold,currency 1,-2301,0,-1839.0,NaN,NaN,0,1642500.0,0.0,NaN,0.0,Car loan,-1836,NaN,0.0,462.0
1715907,438184,5054951,Sold,currency 1,-1284,0,-188.0,NaN,NaN,0,675000.0,0.0,NaN,0.0,Consumer credit,-1151,NaN,0.0,1096.0


### Feature Sold
- Total sold
- Sold finished
- Sold last 90
- Sold last 365

### Feature Bad-Debt
- Total Bad debt
- Bad debt finished
- Sum credit debt
- Sum bad debt overdue
- Avg duration

In [27]:
bureau_bad_debt = bureau[bureau['CREDIT_ACTIVE'] == 'Bad debt']
bureau_bad_debt

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET,DURATION
7778,436084,5723864,Bad debt,currency 1,-986,0,365243.0,NaN,184500.000,0,184500.00,0.000,0.000,0.00,Credit card,-22,NaN,1.0,366229.0
30959,396333,5801236,Bad debt,currency 1,-2627,0,-1529.0,-1502.0,127866.915,0,225000.00,115349.715,-25349.715,0.00,Credit card,-1502,NaN,0.0,1098.0
69946,386859,5246470,Bad debt,currency 1,-2386,26,-2296.0,NaN,53532.000,1,58059.00,NaN,NaN,53532.00,Consumer credit,-2296,NaN,1.0,90.0
295159,231185,5173404,Bad debt,currency 1,-2740,1761,-2558.0,NaN,2137.500,0,14296.50,7717.500,0.000,7717.50,Consumer credit,-888,NaN,0.0,182.0
544624,446651,6439586,Bad debt,currency 1,-2830,0,-2434.0,-1781.0,21741.615,0,58567.68,0.000,0.000,0.00,Consumer credit,-1781,NaN,0.0,396.0
546312,232061,6441729,Bad debt,currency 1,-2899,0,-1773.0,-1434.0,242637.480,0,365670.72,0.000,0.000,0.00,Consumer credit,-1434,NaN,0.0,1126.0
550115,243877,6446445,Bad debt,currency 1,-2493,366,-898.0,NaN,NaN,0,162000.00,NaN,NaN,17541.00,Credit card,-1449,NaN,0.0,1595.0
684142,273612,5309530,Bad debt,currency 1,-2112,0,-1900.0,-1550.0,40871.880,0,22510.89,0.000,0.000,0.00,Consumer credit,-1550,NaN,0.0,212.0
711352,186360,5499851,Bad debt,currency 1,-1218,366,-852.0,NaN,NaN,0,121275.00,0.000,NaN,0.00,Consumer credit,-1035,NaN,0.0,366.0
858406,158069,6039562,Bad debt,currency 1,-1683,366,-862.0,NaN,NaN,0,528165.00,NaN,NaN,503392.50,Consumer credit,-1217,NaN,0.0,821.0


In [28]:
bureau_sold[~bureau_sold["DAYS_ENDDATE_FACT"].isnull()]

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET,DURATION
541,161678,5715165,Sold,currency 1,-1319,0,-1134.0,-75.0,133308.81,0,202500.000,0.0,0.000,0.0,Consumer credit,-71,NaN,1.0,185.0
2493,351664,5717545,Sold,currency 1,-1911,0,-815.0,-134.0,42988.50,0,222750.000,NaN,NaN,0.0,Consumer credit,-22,NaN,1.0,1096.0
5774,161224,5721401,Sold,currency 1,-2777,0,-2474.0,-2085.0,NaN,0,63670.500,0.0,0.000,0.0,Consumer credit,-2076,NaN,0.0,303.0
6347,397627,5722100,Sold,currency 1,-2401,2050,-2031.0,-1726.0,NaN,0,0.000,0.0,0.000,0.0,Credit card,-1724,NaN,0.0,370.0
6356,397627,5722111,Sold,currency 1,-2398,0,-1788.0,-1788.0,0.00,1,230421.645,0.0,190146.645,0.0,Credit card,-1509,NaN,0.0,610.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1712176,194726,5063529,Sold,currency 1,-875,0,220.0,-90.0,0.00,0,1125000.000,0.0,0.000,0.0,Consumer credit,-90,NaN,0.0,1095.0
1713941,205991,5092696,Sold,currency 1,-1297,0,-201.0,-1050.0,0.00,0,337500.000,0.0,0.000,0.0,Others,-1050,NaN,0.0,1096.0
1714121,450700,5066029,Sold,currency 1,-381,0,365243.0,-111.0,0.00,0,45000.000,0.0,0.000,0.0,Credit card,-111,NaN,0.0,365624.0
1714268,107312,5079711,Sold,currency 1,-2207,0,365243.0,-79.0,6597.00,0,108000.000,0.0,0.000,0.0,Credit card,-79,NaN,0.0,367450.0


In [29]:
bureau[bureau['SK_ID_CURR'] == 370974]

,SK_ID_CURR,SK_ID_BUREAU,CREDIT_ACTIVE,CREDIT_CURRENCY,DAYS_CREDIT,CREDIT_DAY_OVERDUE,DAYS_CREDIT_ENDDATE,DAYS_ENDDATE_FACT,AMT_CREDIT_MAX_OVERDUE,CNT_CREDIT_PROLONG,AMT_CREDIT_SUM,AMT_CREDIT_SUM_DEBT,AMT_CREDIT_SUM_LIMIT,AMT_CREDIT_SUM_OVERDUE,CREDIT_TYPE,DAYS_CREDIT_UPDATE,AMT_ANNUITY,TARGET,DURATION
460816,370974,5043240,Active,currency 1,-494,0,31053.0,NaN,NaN,0,477000.0,456813.0,NaN,0.0,Credit card,-10,NaN,0.0,31547.0
460817,370974,5043241,Active,currency 1,-518,0,31053.0,NaN,NaN,0,675000.0,607828.5,NaN,0.0,Credit card,-10,NaN,0.0,31571.0
460818,370974,5043243,Closed,currency 1,-2564,0,-2381.0,-2410.0,0.0,0,25987.5,0.0,0.0,0.0,Consumer credit,-1713,NaN,0.0,183.0
460819,370974,5043244,Active,currency 1,-179,0,1650.0,NaN,0.0,0,1350000.0,1304289.0,0.0,0.0,Consumer credit,-19,NaN,0.0,1829.0
1715601,370974,5043237,Active,currency 1,-179,0,31053.0,NaN,NaN,0,0.0,0.0,0.0,0.0,Credit card,-24,NaN,0.0,31232.0
1715602,370974,5043238,Sold,currency 1,-2301,0,-1839.0,NaN,NaN,0,1642500.0,0.0,NaN,0.0,Car loan,-1836,NaN,0.0,462.0
